reactive/azure.py

```diff
@@ -44,6 +45,11 @@ def no_requests():
     layer.status.active('ready')
 
 
+@when_any('config.changed')
+def rehandle_requests():
+    handle_requests()
```

Comment:
This doesn't look quite right to me either. The main problem is that even if handle_requests runs again, it's only going to iterate over and handle requests that have changed (see [here](https://github.com/juju-solutions/interface-azure-integration/blob/ffdc0866d860517a0fee54322b15004193f13041/provides.py#L55-L66)). It won't re-handle existing requests.

The good news is that the openstack-integrator charm does this properly and I think we should follow its example. Here is what I would recommend:

1. Replace [AzureIntegrationProvides.requests](https://github.com/juju-solutions/interface-azure-integration/blob/ffdc0866d860517a0fee54322b15004193f13041/provides.py#L56) with all_requests and new_requests properties. ([Example](https://github.com/charmed-kubernetes/interface-openstack-integration/blob/0d7a994f04b9e92ed847829ce8349b1a9c672e47/provides.py#L44-L60))
2. Use `@when_any('endpoint.clients.requests-pending', 'config.changed')` to make it so handle_requests runs when requests change *or* when config changes. ([Example](https://github.com/charmed-kubernetes/charm-openstack-integrator/blob/53fe9535f9e1e092e3ff207a18ce59350f5d9219/reactive/openstack.py#L64-L65))
3. Update handle_requests to use all_requests if config has changed, new_requests if it has not. ([Example](https://github.com/charmed-kubernetes/charm-openstack-integrator/blob/53fe9535f9e1e092e3ff207a18ce59350f5d9219/reactive/openstack.py#L83))
4. Update [AzureIntegrationRequires.check_ready](https://github.com/juju-solutions/interface-azure-integration/blob/ffdc0866d860517a0fee54322b15004193f13041/requires.py#L108) to also set a `ready.changed` flag if cloud data has changed. ([Example 1](https://github.com/charmed-kubernetes/interface-openstack-integration/blob/0d7a994f04b9e92ed847829ce8349b1a9c672e47/requires.py#L80-L81), [Example 2](https://github.com/charmed-kubernetes/interface-openstack-integration/blob/0d7a994f04b9e92ed847829ce8349b1a9c672e47/requires.py#L104-L122))
5. Update kubernetes-master [update_cloud_config](https://github.com/charmed-kubernetes/charm-kubernetes-master/blob/aab20cac766902b6e3957c7b9e4a09af1bcbde3b/reactive/kubernetes_master.py#L2692-L2707) to also handle `endpoint.azure.ready.changed`. It should clear the `kubernetes-master.cloud.ready` and `endpoint.vsphere.ready.changed` flags. This is needed to ensure the cloud configs get re-rendered by the [cloud_ready](https://github.com/charmed-kubernetes/charm-kubernetes-master/blob/aab20cac766902b6e3957c7b9e4a09af1bcbde3b/reactive/kubernetes_master.py#L2670-L2689) handler.

Reply:
I'm okay with deferring the config.changed handling to another bug/PR.

But before we merge this PR, I would request two minor changes:
1. Remove the `rehandle_requests` handler. As it is now, this introduces a path where `handle_requests()` might get called even though flags like `apt.installed.azure-cli` and `charm.azure.creds.set` are not set. That's dangerous and could lead to errors.
2. Update the config descriptions for vnetName, vnetResourceGroup, subnetName, and vnetSecurityGroup to make it clear that these cannot be changed. Something like "This config must be set at deployment and cannot be changed later."